python-sdk/nuscenes/eval/predict/metrics.py

```diff
@@ -0,0 +1,436 @@
+        results = miss_rate_top_k(prediction.prediction, ground_truth,
+                                 prediction.probabilities, self.tolerance)
+        return desired_number_of_modes(results, self.k_to_report)
+
+    def serialize(self) -> Dict[str, Any]:
+        return {'k_to_report': self.k_to_report,
+                'name': 'MissRateTopK',
+                'aggregators': [agg.serialize() for agg in self.aggregators],
+                'tolerance': self.tolerance}
+
+    @property
+    def aggregators(self,) -> List[Aggregator]:
+        return self._aggregators
+
+    @property
+    def name(self):
+        return f"MissRateTopK_{self.tolerance}"
+
+    @property
+    def shape(self):
+        return len(self.k_to_report)
+
+
+class OffRoadRate(Metric):
+
+    def __init__(self, helper: PredictHelper, aggregators: List[Aggregator]):
+        self._aggregators = aggregators
+        self.helper = helper
+        self.drivable_area_polygons = self.load_drivable_area_masks(helper)
+        self.pixels_per_meter = 10
+
+    @staticmethod
+    def load_drivable_area_masks(helper: PredictHelper) -> Dict[str, np.ndarray]:
+        """
+        Loads the polygon representation of the drivable area for each map
+        :param helper: Instance of PredictHelper
+        :return: Mapping from map_name to drivable area polygon
+        """
+
+        maps: Dict[str, NuScenesMap] = load_all_maps(helper)
+
+        masks = {}
+        for map_name, map_api in maps.items():
+
+            masks[map_name] = map_api.get_map_mask(patch_box=None, patch_angle=0, layer_names=['drivable_area'],
+                                                   canvas_size=None)[0]
+
+        return masks
+
+    @staticmethod
+    def interpolate_path(mode: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
+        """
+        Interpolate trajectory with a cubic spline if there are enough points.
+        """
+
+        # interpolate.splprep needs unique points.
+        # We use a loop as opposed to np.unique because
+        # the order of the points must be the same
+        seen = set()
+        ordered_array = []
+        for row in mode:
+            row_tuple = tuple(row)
+            if row_tuple not in seen:
+                seen.add(row_tuple)
+                ordered_array.append(row_tuple)
+
+        new_array = np.array(ordered_array)
+
+        unique_points = np.atleast_2d(new_array)
+
+        if unique_points.shape[0] <= 3:
+            return unique_points[:, 0], unique_points[:, 1]
+        else:
+            knots, _ = interpolate.splprep([unique_points[:, 0], unique_points[:, 1]], k=3, s=0.1)
+            x_interpolated, y_interpolated = interpolate.splev(np.linspace(0, 1, 200), knots)
+            return x_interpolated, y_interpolated
+
+    def __call__(self, ground_truth: np.ndarray, prediction: Prediction) -> np.ndarray:
+        map_name = self.helper.get_map_name_from_sample_token(prediction.sample)
```

Comment:
An explanation of OffRoadRate may be helpful here.

Reply:
Added!